driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java

```diff
@@ -102,37 +107,43 @@
             this.operationTime = result.getTimestamp(OPERATION_TIME, null);
             this.postBatchResumeToken = getPostBatchResumeTokenFromResponse(result);
         }
-        if (firstQueryResult.getCursor() != null) {
+        ServerCursor serverCursor = initFromQueryResult(firstQueryResult);
+        if (serverCursor != null) {
             notNull("connectionSource", connectionSource);
         }
-        if (connectionSource != null) {
-            this.connectionSource = connectionSource.retain();
-        }
-
-        initFromQueryResult(firstQueryResult);
         firstBatchEmpty = firstQueryResult.getResults().isEmpty();
-
+        Connection connectionToPin = null;
+        boolean releaseServerAndResources = false;
         if (connection != null) {
             this.maxWireVersion = connection.getDescription().getMaxWireVersion();
             if (limitReached()) {
-                killCursor(connection);
+                releaseServerAndResources = true;
             } else {
                 assertNotNull(connectionSource);
                 if (connectionSource.getServerDescription().getType() == ServerType.LOAD_BALANCER) {
-                    this.connection = connection.retain();
-                    this.connection.markAsPinned(Connection.PinningMode.CURSOR);
+                    connectionToPin = connection;
                 }
             }
         }
-        releaseConnectionAndSourceIfNoServerCursor();
+        closeableState = new CloseableState(namespace, connectionSource, connectionToPin, serverCursor);
+        if (releaseServerAndResources) {
+            closeableState.releaseServerAndClientResources(assertNotNull(connection));
+        }
     }
 
     @Override
     public boolean hasNext() {
-        if (closed) {
+        if (!closeableState.startOperation()) {
```

Comment:
I think the naming here could be worked on. It reads if not started the cursor has been closed. Which doesn't make sense.